segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -717,57 +727,49 @@ public void readTableKeys(WireCommands.ReadTableKeys readTableKeys) {
+        val continuationToken = new AtomicReference<>(EMPTY_BUFFER);
+        val keys = new ArrayList<WireCommands.TableKey>();
         val timer = new Timer();
         tableStore.keyIterator(segment, state, TIMEOUT)
-                  .thenCompose(itr -> itr.collectRemaining(
-                          e -> {
-                              synchronized (keys) {
-                                  if (keys.size() < suggestedKeyCount && msgSize.get() < MAX_READ_SIZE) {
-                                      Collection<TableKey> tableKeys = e.getEntries();
-                                      ArrayView lastState = e.getState();
-
-                                      // Store all tableKeys.
-                                      keys.addAll(tableKeys);
-                                      // update the continuation token.
-                                      continuationToken.set(wrappedBuffer(lastState.array(), lastState.arrayOffset(), lastState.getLength()));
-                                      // Update msgSize.
-                                      msgSize.addAndGet(getTableKeyBytes(segment, tableKeys, lastState.getLength()));
-                                      return true;
-                                  } else {
-                                      return false;
-                                  }
-                              }
-                          }))
-                  .thenAccept(v -> {
-                      final List<WireCommands.TableKey> wireCommandKeys;
-                      synchronized (keys) {
-                          log.debug(readTableKeys.getRequestId(), "{} keys obtained for ReadTableKeys request.", keys.size());
-                          wireCommandKeys = keys.stream()
-                                                .map(k -> {
-                                                    ArrayView keyArray = k.getKey();
-                                                    return new WireCommands.TableKey(wrappedBuffer(keyArray.array(),
-                                                                                                   keyArray.arrayOffset(),
-                                                                                                   keyArray.getLength()), k.getVersion());
-                                                })
-                                                .collect(toList());
-                      }
-                      connection.send(new WireCommands.TableKeysRead(readTableKeys.getRequestId(), segment, wireCommandKeys, continuationToken.get()));
-                      this.tableStatsRecorder.iterateKeys(readTableKeys.getSegment(), keys.size(), timer.getElapsed());
-                  }).exceptionally(e -> handleException(readTableKeys.getRequestId(), segment, operation, e));
+                .thenCompose(itr -> itr.collectRemaining(
+                        e -> {
```

Comment:
Why the synchronization blocks is not needed anymore here?

Reply:
There was nothing to synchronize, since these callbacks would be invoked in order. But thanks for pointing this out as I had to make `keys` a synchronized list to make it safe for passing between threads.

Reply:
There is no race here. `itr.collectRemaining(...)` will fully execute before `thenAccept` (count the number of open and close parens). So the iterator will be fully consumed (collected) before we invoke `connection.send`. 

In this case, `Collection.synchronizedList` is the appropriate solution as all I need to worry about is passing data between threads, but at no given time will I both iterate and update that collection.

Reply:
After talking offline, I understand what the issue is. The `Collections.synchronizedList` returns a wrapper that synchronizes every access except the one to the `iterator()` which is delegated to the wrapped List, so that one is not thread safe.

I am bringing back the `synchronized` block.